File: util/stage/Stage.java
package common.util.stage;

import common.CommonStatic;
import common.io.assets.Admin.StaticPermitted;
import common.io.json.Dependency;
import common.io.json.JsonClass;
import common.io.json.JsonClass.NoTag;
import common.io.json.JsonField;
import common.pack.Identifier;
import common.pack.IndexContainer;
import common.pack.PackData.PackDesc;
import common.pack.Source.ResourceLocation;
import common.pack.UserProfile;
import common.system.BasedCopable;
import common.system.files.VFile;
import common.util.BattleStatic;
import common.util.Data;
import common.util.lang.MultiLangCont;
import common.util.lang.MultiLangData;
import common.util.pack.Background;
import common.util.stage.SCDef.Line;
import common.util.stage.info.DefStageInfo;
import common.util.stage.info.StageInfo;
import common.util.unit.Enemy;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.*;

@IndexContainer.IndexCont(StageMap.class)
@JsonClass.JCGeneric(Identifier.class)
@JsonClass(noTag = NoTag.LOAD)
public class Stage extends Data
		implements BasedCopable<Stage, StageMap>, BattleStatic, IndexContainer.Indexable<StageMap, Stage> {

	@StaticPermitted
	public static final MapColc CLIPMC = new MapColc.ClipMapColc();
	@StaticPermitted
	public static final StageMap CLIPSM = CLIPMC.maps.get(0);
	@StaticPermitted
	private static final int[] CH_CASTLES = { 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28,
			27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 46,
			47, 45, 47, 47, 45, 45 };

	public StageInfo info;
	@JsonField(block = true)
	public boolean isBCstage = false;

	@JsonClass.JCIdentifier
	public final Identifier<Stage> id;

	@JsonField(io = JsonField.IOType.R)
	public String name = "";
	@JsonField(generic = MultiLangData.class)
	public MultiLangData names = new MultiLangData();

	public boolean non_con, trail;
	public int len, health, max, mush, bgh;
	public int timeLimit = 0;
	public int minSpawn = 1, maxSpawn = 1;
	public Identifier<CastleImg> castle;
	public Identifier<Background> bg, bg1;
	public Identifier<Music> mus0, mus1;
	public SCDef data;
	public Limit lim;
	@JsonField(generic = Replay.class, alias = ResourceLocation.class)
	public ArrayList<Replay> recd = new ArrayList<>();

	@JsonClass.JCConstructor
	public Stage() {
		id = null;
	}

	public Stage(Identifier<Stage> id) {
		this.id = id;
		len = 3000;
		health = 60000;
		max = 8;
		names.put("stage " + getCont().list.size());
		lim = new Limit();
		data = new SCDef(0);
	}

	public Stage(StageMap sm) {
		this.id = sm.getNextID();
		len = 3000;
		health = 60000;
		max = 8;
		names.put("stage " + sm.list.size());
		lim = new Limit();
		data = new SCDef(0);
	}

	@SuppressWarnings("deprecation")
	protected Stage(Identifier<Stage> id, VFile f, int type) {
		this.id = id;
		isBCstage = true;
		StageMap sm = getCont();
		if (sm.info != null)
			sm.info.getData(this);
		Queue<String> qs = f.getData().readLine();
		names.put("" + id);
		String temp;
		if (type == 0) {
			temp = qs.poll();
			String[] strs = temp.split(",");
			int cas = CommonStatic.parseIntN(strs[0]);
			if (cas == -1)
				cas = CH_CASTLES[id.id];
			if (sm.cast != -1)
				cas = sm.cast * 1000 + cas;
			castle = Identifier.parseInt(cas, CastleImg.class);
			non_con = strs[1].equals("1");

			if(info != null)
				((DefStageInfo)info).setData(strs);
		} else {
			castle = Identifier.parseInt(sm.cast * 1000 + CH_CASTLES[id.id], CastleImg.class);
			non_con = false;
		}
		int intl = type == 2 ? 9 : 10;
		String[] strs = qs.poll().split(",");
		len = Integer.parseInt(strs[0]);
		health = Integer.parseInt(strs[1]);
		minSpawn = Integer.parseInt(strs[2]);
		maxSpawn = Integer.parseInt(strs[3]);
		bg = Identifier.rawParseInt(Integer.parseInt(strs[4]), Background.class);
		max = Integer.parseInt(strs[5]);
		timeLimit = strs.length >= 8 ? Math.max(Integer.parseInt(strs[7]), 0) : 0;
		if(timeLimit != 0)
			health = Integer.MAX_VALUE;
		trail = timeLimit != 0;
		int isBase = Integer.parseInt(strs[6]) - 2;
		List<int[]> ll = new ArrayList<>();

		while (qs.size() > 0)
			if ((temp = qs.poll()).length() > 0) {
				if (!Character.isDigit(temp.charAt(0)))
					break;
				if (temp.startsWith("0,"))
					break;
				String[] ss = temp.split(",");
				int[] data = new int[SCDef.SIZE];
				for (int i = 0; i < intl; i++)
					if(i < ss.length)
						data[i] = Integer.parseInt(ss[i]);
					else
						//Handle missing value manually
						if(i == 9)
							data[i] = 100;
				data[0] -= 2;
				data[2] *= 2;
				data[3] *= 2;
				data[4] *= 2;
				if (timeLimit == 0 && intl > 9 && data[5] > 100 && data[9] == 100) {
					data[9] = data[5];
					data[5] = 100;
				}
				if (ss.length > 11 && CommonStatic.isInteger(ss[11]))
					data[SCDef.M1] = Integer.parseInt(ss[11]);
				else
					data[SCDef.M1] = data[SCDef.M];

				if (data[0] == isBase)
					data[SCDef.C0] = 0;
				ll.add(data);
			}
		SCDef scd = new SCDef(ll.size());
		for (int i = 0; i < ll.size(); i++)
			scd.datas[i] = new Line(ll.get(scd.datas.length - i - 1));

		int ano = CommonStatic.parseIntN(strs[6]);
		if (ano == 317)
			scd.datas[ll.size() - 1].castle_0 = 0;
		data = scd;
		validate();
	}

	public boolean contains(Enemy e) {
		return data.contains(e);
	}

	@Override
	public Stage copy(StageMap sm) {
		Stage ans = new Stage(sm);
		ans.len = len;
		ans.health = health;
		ans.max = max;
		if (bg != null)
			ans.bg = bg.clone();
		if (bg1 != null)
			ans.bg1 = bg1.clone();
		if (castle != null)
			ans.castle = castle.clone();
		ans.names.put(toString());
		ans.data = data.copy();
		ans.lim = lim != null ? lim.clone() : getLim(0);
		ans.non_con = non_con;
		if (mus0 != null)
			ans.mus0 = mus0.clone();
		if (mus1 != null)
			ans.mus1 = mus1.clone();
		ans.bgh = bgh;
		ans.mush = mush;
		return ans;
	}

	@Override
	public Identifier<Stage> getID() {
		return id;
	}

	public Limit getLim(int star) {
		Limit tl = new Limit();
		if (lim != null && (lim.star == -1 || lim.star == star))
			tl.combine(lim);
		for (Limit l : getCont().lim)
			if (l.star == -1 || l.star == star)
				if (l.sid == -1 || l.sid == id())
					tl.combine(l);
		return tl;
	}

	public int id() {
		return getCont().list.indexOf(this);
	}

	public Set<String> isSuitable(String pack) {
		Dependency dep = Dependency.collect(this);
		PackDesc desc = UserProfile.getUserPack(pack).desc;
		Set<String> set = dep.getPacks();
		System.out.println("Stage: " + set);// FIXME
		set.remove(Identifier.DEF);
		set.remove(pack);
		for (String str : desc.dependency)
			set.remove(str);
		return set;
	}

	public void setNames(String str) {
		while (!checkName(str))
			str += "'";
		names.put(str);
	}

	@Override
	public String toString() {
		String desp = MultiLangCont.get(this);
		if (desp != null && desp.length() > 0)
			return desp;
		String n = names.toString();
		if (n.length() > 0)
			return n;
		return map + " - " + id();
	}

	public boolean isAkuStage() {
		if(data.datas.length == 0)
			return false;

		Line line = data.datas[data.datas.length - 1];

		if(line.enemy == null)
			return false;

		return line.enemy.id == 574 && line.enemy.pack.equals(Identifier.DEF) && line.castle_0 == 0;
	}

	protected void validate() {
		if(trail)
			return;

		if(getCont() == null || getCont().getCont() == null)
			return;

		if(getCont().getCont().getSID().equals("000006") || getCont().getCont().getSID().equals("000011"))
			trail = data.isTrail();
	}

	private boolean checkName(String str) {
		for (Stage st : getCont().list)
			if (st != this && st.names.toString().equals(str))
				return false;
		return true;
	}
}

File: util/stage/Stage.kt
package common.util.stage

import common.CommonStatic
import common.io.InStream
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonClass
import common.io.json.JsonEncoder
import common.io.json.JsonField
import common.io.json.JsonField.GenType
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.PackData
import common.pack.PackData.UserPack
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.pack.VerFixer.VerFixerException
import common.system.BasedCopable
import common.system.files.VFile
import common.util.BattleStatic
import common.util.Data
import common.util.lang.MultiLangCont
import common.util.pack.Background
import common.util.stage.EStage
import common.util.stage.Limit.PackLimit
import common.util.stage.MapColc.ClipMapColc
import common.util.stage.Recd
import common.util.stage.SCDef
import common.util.stage.Stage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.Enemy
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.util.*

@JsonClass
class Stage : Data, BasedCopable<Stage?, StageMap?>, BattleStatic {
    class StageInfo(info: StageMapInfo, s: Stage, data: IntArray) {
        val st: Stage
        val map: StageMapInfo
        val energy: Int
        val xp: Int
        val once: Int
        var rand = 0
        val drop: Array<IntArray?>
        val time: Array<IntArray>
        var diff = -1
        fun getHTML(): String {
            var ans = "<html>energy cost: $energy<br> xp: $xp<br> drop rewards: "
            if (drop.size == 0) ans += "none" else if (drop.size == 1) ans += ("{chance: " + drop[0]!![0] + "%, item ID: " + drop[0]!![1] + ", number: " + drop[0]!![2] + "}, once: "
                    + once) else {
                ans += "count: " + drop.size + ", rand mode: " + rand + ", once: " + once + "<br>"
                ans += "<table><tr><th>chance</th><th>item ID</th><th>number</th></tr>"
                for (dp in drop) ans += "<tr><td>" + dp!![0] + "%</td><td>" + dp[1] + "</td><td>" + dp[2] + "</td><tr>"
                ans += "</table>"
            }
            if (time.size > 0) {
                ans += "<br> time scores: count: " + time.size + "<br>"
                ans += "<table><tr><th>score</th><th>item ID</th><th>number</th></tr>"
                for (tm in time) ans += "<tr><td>" + tm[0] + "</td><td>" + tm[1] + "</td><td>" + tm[2] + "</td><tr>"
                ans += "</table>"
            }
            return ans
        }

        init {
            map = info
            st = s
            energy = data[0]
            xp = data[1]
            s.mus0 = PackData.Identifier.Companion.parseInt<Music>(data[2], Music::class.java)
            s.mush = data[3]
            s.mus1 = PackData.Identifier.Companion.parseInt<Music>(data[4], Music::class.java)
            once = data[data.size - 1]
            var isTime = data.size > 15
            if (isTime) for (i in 8..14) if (data[i] != -2) isTime = false
            if (isTime) {
                time = Array((data.size - 17) / 3) { IntArray(3) }
                for (i in time.indices) for (j in 0..2) time[i][j] = data[16 + i * 3 + j]
            } else time = Array(0) { IntArray(3) }
            val isMulti = !isTime && data.size > 9
            if (data.size == 6) {
                drop = arrayOfNulls(0)
                rand = 0
            } else if (!isMulti) {
                drop = arrayOfNulls(1)
                rand = 0
            } else {
                drop = Array((data.size - 7) / 3) { IntArray(3) }
                rand = data[8]
                for (i in 1 until drop.size) for (j in 0..2) drop[i]!![j] = data[6 + i * 3 + j]
            }
            if (drop.size > 0) drop[0] = intArrayOf(data[5], data[6], data[7])
        }
    }

    var info: StageInfo? = null
    var recd: List<Recd> = ArrayList<Recd>()
    val map: StageMap

    @JsonField
    var name = ""

    @JsonField
    var non_con = false

    @JsonField
    var trail = false

    @JsonField
    var len: Int

    @JsonField
    var health: Int

    @JsonField
    var max: Int

    @JsonField
    var mush = 0
    var castle: PackData.Identifier<CastleImg>? = null
    var bg: PackData.Identifier<Background>? = null
    var mus0: PackData.Identifier<Music>? = null
    var mus1: PackData.Identifier<Music>? = null

    @JsonField
    var loop0: Long = 0

    @JsonField
    var loop1: Long = 0

    @JsonField
    var data: SCDef?

    @JsonField(gen = GenType.GEN)
    var lim: Limit? = null

    constructor(sm: StageMap) {
        map = sm
        len = 3000
        health = 60000
        max = 8
        name = "stage " + sm.list.size
        lim = Limit()
        data = SCDef(0)
    }

    @Deprecated("")
    constructor(pack: UserPack?, sm: StageMap, `is`: InStream) : this(sm) {
        val `val`: Int = Data.Companion.getVer(`is`.nextString())
        if (`val` != 409) throw VerFixerException("stage version has to be 409, got$`val`")
        name = `is`.nextString()
        bg = PackData.Identifier.Companion.parseInt<Background>(`is`.nextInt(), Background::class.java)
        castle = PackData.Identifier.Companion.parseInt<CastleImg>(`is`.nextInt(), CastleImg::class.java)
        health = `is`.nextInt()
        len = `is`.nextInt()
        mus0 = PackData.Identifier.Companion.parseInt<Music>(`is`.nextInt(), Music::class.java)
        mush = `is`.nextInt()
        mus1 = PackData.Identifier.Companion.parseInt<Music>(`is`.nextInt(), Music::class.java)
        loop0 = `is`.nextLong()
        loop1 = `is`.nextLong()
        max = `is`.nextByte()
        non_con = `is`.nextByte() == 1
        data = SCDef.Companion.zread(`is`.subStream())
        lim = PackLimit(pack, `is`)
        val t: Int = `is`.nextInt()
        for (i in 0 until t) {
            val name: String = `is`.nextString()
            Recd.Companion.getRecd(this, `is`.subStream(), name)
        }
        validate()
    }

    constructor(sm: StageMap, id: Int, f: VFile<*>, type: Int) {
        map = sm
        if (sm.info != null) sm.info.getData(this)
        val qs: Queue<String> = f.getData().readLine()
        name = "" + id
        var temp: String
        if (type == 0) {
            temp = qs.poll()
            val strs = temp.split(",").toTypedArray()
            var cas: Int = CommonStatic.parseIntN(strs[0])
            if (map.cast != -1) cas += map.cast * 1000
            castle = PackData.Identifier.Companion.parseInt<CastleImg>(cas, CastleImg::class.java)
            non_con = strs[1] == "1"
        } else {
            castle = PackData.Identifier.Companion.parseInt<CastleImg>(0, CastleImg::class.java)
            non_con = false
        }
        val intl = if (type == 2) 9 else 10
        val strs = qs.poll().split(",").toTypedArray()
        len = strs[0].toInt()
        health = strs[1].toInt()
        bg = PackData.Identifier.Companion.parseInt<Background>(strs[4].toInt(), Background::class.java)
        max = strs[5].toInt()
        val isBase = strs[6].toInt() - 2
        val ll: MutableList<IntArray> = ArrayList()
        while (qs.size > 0) if (qs.poll().also { temp = it }.length > 0) {
            if (!Character.isDigit(temp[0])) break
            if (temp.startsWith("0,")) break
            val ss = temp.split(",").toTypedArray()
            val data = IntArray(SCDef.Companion.SIZE)
            for (i in 0 until intl) data[i] = ss[i].toInt()
            data[0] -= 2
            data[2] *= 2
            data[3] *= 2
            data[4] *= 2
            if (intl > 9 && data[5] > 100 && data[9] == 100) {
                data[9] = data[5]
                data[5] = 100
            }
            if (ss.size > 11 && CommonStatic.isInteger(ss[11])) data[SCDef.Companion.M1] = ss[11].toInt() else data[SCDef.Companion.M1] = data[SCDef.Companion.M]
            if (data[0] == isBase) data[SCDef.Companion.C0] = 0
            ll.add(data)
        }
        val scd = SCDef(ll.size)
        for (i in ll.indices) scd.datas.get(i) = SCDef.Line(ll[scd.datas.size - i - 1])
        if (strs.size > 6) {
            val ano: Int = CommonStatic.parseIntN(strs[6])
            if (ano == 317) scd.datas.get(ll.size - 1)!!.castle_0 = 0
        }
        data = scd
        validate()
    }

    operator fun contains(e: Enemy): Boolean {
        return data!!.contains(e)
    }

    override fun copy(sm: StageMap): Stage {
        val ans = Stage(sm)
        ans.len = len
        ans.health = health
        ans.max = max
        ans.bg = bg!!.clone()
        ans.castle = castle!!.clone()
        ans.name = name
        ans.data = data!!.copy()
        if (lim != null) ans.lim = lim!!.clone()
        ans.mus0 = mus0!!.clone()
        ans.mus1 = mus1!!.clone()
        ans.mush = mush
        return ans
    }

    fun getLim(star: Int): Limit {
        val tl = Limit()
        if (lim != null && (lim!!.star == -1 || lim!!.star == star)) tl.combine(lim)
        for (l in map.lim) if (l.star == -1 || l.star == star) if (l.sid == -1 || l.sid == id()) tl.combine(l)
        return tl
    }

    fun id(): Int {
        return map.list.indexOf(this)
    }

    fun isSuitable(pack: String?): Boolean {
        // FIXME suitable
        return false
    }

    fun setName(str: String) {
        // TODO validate str = MainBCU.validate(str);
        var str = str
        while (!checkName(str)) str += "'"
        name = str
    }

    override fun toString(): String {
        val desp: String = MultiLangCont.Companion.get(this)
        if (desp != null && desp.length > 0) return desp
        return if (name.length > 0) name else map.toString() + " - " + id()
    }

    fun validate() {
        trail = data!!.isTrail()
    }

    private fun checkName(str: String): Boolean {
        for (st in map.list) if (st !== this && st.name == str) return false
        return true
    }

    companion object {
        val CLIPMC: MapColc = ClipMapColc()
        val CLIPSM: StageMap = Stage.Companion.CLIPMC.maps.get(0)
    }
}
